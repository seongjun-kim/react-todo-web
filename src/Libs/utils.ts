import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};
