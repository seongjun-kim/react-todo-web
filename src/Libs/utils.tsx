import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, BrowserRouter } from 'react-router-dom';
import { ToDoListProvider } from 'Contexts';
import App from 'App';

/**
 * [renderWithRouter]
 * useLocation과 같은 라우터 관련 훅을 포함한 단일 컴포넌트 테스트를 위한 렌더링 함수
 * 컴포넌트 테스트를 위한 함수이므로 Components 폴더 아래 테스트에 활용
 */
export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};

/**
 * [renderWithMemoryRouter]
 * 실제 앱에서의 구동 방식 테스트를 위한 렌더링 함수
 * 페이지 전환 테스트를 위한 함수이므로 Pages 폴더 아래 테스트에 활용
 *
 *  - initialEntries를 window.history 처럼 활용 가능
 *  - TestComponent를 활용한 앱 내 추가 컴포넌트 적용 테스트 가능
 *    - 주로 useLocation().pathname을 파악하는 용도로 활용
 */
export const renderWithMemoryRouter = (
  { initialIndex = 0, initialEntries = ['/'] } = {},
  TestComponent?: JSX.Element | JSX.Element[],
) => {
  return render(
    <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
      {TestComponent}
      <ToDoListProvider>
        <App />
      </ToDoListProvider>
    </MemoryRouter>,
  );
};
