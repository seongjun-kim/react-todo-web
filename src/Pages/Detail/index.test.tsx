import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import 'jest-styled-components';
import { renderWithMemoryRouter } from 'Libs/utils';
import { useLocation } from 'react-router-dom';

describe('<Detail />', () => {
  const sampleData = ['ToDo 1'];
  localStorage.setItem('ToDoList', JSON.stringify(sampleData));
  it('renders component correctly', () => {
    const { container } = renderWithMemoryRouter({
      initialEntries: ['/detail/0'],
    });

    const toDoItem = screen.getByText(sampleData[0]);
    expect(toDoItem).toBeInTheDocument();

    const button = screen.getByText('Delete');
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('renders nothing without id', () => {
    renderWithMemoryRouter({ initialEntries: ['/detail'] });

    expect(screen.queryAllByText(sampleData[0]).length).toBe(0);
    expect(screen.queryAllByText('Delete').length).toBe(0);
  });

  it('deletes ToDo data', () => {
    const TestComponent = () => {
      const { pathname } = useLocation();
      return <div>{pathname}</div>;
    };
    renderWithMemoryRouter(
      {
        initialEntries: ['/', '/detail/0'],
        initialIndex: 1,
      },
      <TestComponent />,
    );

    const url = screen.getByText('/detail/0');
    expect(url).toBeInTheDocument();

    const toDoItem = screen.getByText('ToDo 1');
    expect(toDoItem).toBeInTheDocument();
    const button = screen.getByText('Delete');
    expect(button).toBeInTheDocument();

    fireEvent.click(button);

    expect(url.textContent).toBe('/');
    expect(toDoItem).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();
  });
});
