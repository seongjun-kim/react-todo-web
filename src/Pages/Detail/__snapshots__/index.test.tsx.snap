// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> renders component correctly 1`] = `
.c6 {
  text-align: center;
  background-color: #ff1744;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c6:hover {
  background-color: #f01440;
}

.c6:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c7 {
  color: #fff;
  font-size: 16px;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #1e40ff;
}

.c2 {
  padding: 20px;
  color: #fff;
  font-size: 20px;
  font-weight: 600;
}

.c3 {
  padding: 20px;
  color: #fff;
  font-size: 16px;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: absolute;
  left: 20px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  min-width: 350px;
  height: 350px;
  overflow-y: auto;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
  padding: 10px;
}

.c0 {
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flexx;
  display: -ms-flexxbox;
  display: flexx;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Todo In Detail
      </div>
      <a
        class="c3"
        href="/"
      >
        Go Back...
      </a>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        ToDo 1
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          Delete
        </div>
      </div>
    </div>
  </div>
</div>
`;
